File: Trees/isTreeSymmetric.js
// Given a binary tree t, determine whether it is symmetric 
// around its center, i.e.each side mirrors the other.

function isTreeSymmetric(t){
    if( t===null )
        return true;
    if( (t.left === null) || (t.right === null))
        return false;
    let queue = [];
    queue.push(t.left);
    queue.push(t.right);

    while( q.length !== 0){
        tempLeft = q.shift();
        tempRight = q.shift();

        if( tempLeft==null && tempRight==null){
            continue;
        }

        if( (tempLeft==null && tempRight!==null) ||
            (tempRight==null && tempRight!==null) ){
                return false;
        }

        if(tempLeft.value !== tempRight.value){
            return false;
        }
        
        // this order is important
        q.push(tempLeft.left);
        q.push(tempRight.right);
        q.push(tempLeft.right);
        q.push(tempRight.left);
    }
    return true;
}

File: Trees/isTreeSymmetric.py
# Given a binary tree t, determine whether it is symmetric
# around its center, i.e.each side mirrors the other.

def isTreeSymmetric(t):
    if not t:
        return True
    if not t.left or not t.right:
        return False
    queue = []
    queue.append(t.left)
    queue.append(t.right)

    while len(queue) != 0:
        tempLeft = queue.pop(0)
        tempRight = queue.pop(0)

        if not tempLeft and not tempRight:
            continue

        if not tempLeft and not tempLeft or not tempRight and not tempRight:
            return False

        if tempLeft.value != tempRight.value:
            return False

        queue.append(tempLeft.left)
        queue.append(tempRight.right)
        queue.append(tempLeft.right)
        queue.append(tempRight.left)

    return True
